### PR TITLE
feat: add numbers.range_step built-in function

### PR DESCRIPTION
**Pull request by sspaink**
### Why the changes in this PR are needed?

resolves: https://github.com/open-policy-agent/opa/issues/6186

### What are the changes in this PR?

Adds a built-in function to create an array that increments by a user provided step value.

### Notes to assist PR review:

* Decided to restrict the step number to only positive values to allow easy reuse of the existing code.
* Wasn't sure what the numbering convention is for the test files, so I just started at 1.

### CODE REVIEW

File: test/cases/testdata/numbersrangestep/test-numbersrangestep-1.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+cases:
+- data: {}
+  modules:
+  - |
+    package generated
+
+    p = __local0__ {
```
Comment (anderseknert) [inline test/cases/testdata/numbersrangestep/test-numbersrangestep-1.yaml:7]: I presume you've had a look at other examples for these tests, so I don't blame you for copying their style — but many of them were generated from code eons ago :) There's no need to use var names like `__local0__` here. Using normal variable names and identifiers is preferable :)

Also, using a more descriptive package name than `generated` would be good.

**Review by anderseknert — COMMENTED** [review]
Code looks good to me. Nice first contribution!

Just a comment on the tests, but nothing major.

File: test/cases/testdata/numbersrangestep/test-numbersrangestep-1.yaml (hunk 1 above)
Comment (sspaink) [reply test/cases/testdata/numbersrangestep/test-numbersrangestep-1.yaml]: Thank you for the explanation! I was confused why `__local0__` was being used. I've updated the tests as you suggested and to follow [the style outlined by the documentation instead](https://www.openpolicyagent.org/docs/latest/contrib-adding-builtin-functions/#test).

**Review by johanfylling — COMMENTED** [review]
Thank you for your contribution @sspaink! 😃
Just a few questions and comments.

File: topdown/numbers.go [hunk 2]
```diff
@@ -28,32 +28,72 @@ func builtinNumbersRange(bctx BuiltinContext, operands []*ast.Term, iter func(*a
...
-		for i := new(big.Int).Set(x); i.Cmp(y) >= 0; i = i.Sub(i, one) {
-			if bctx.Cancel != nil && bctx.Cancel.Cancelled() {
-				return haltErr
-			}
-			result = result.Append(ast.NewTerm(builtins.IntToNumber(i)))
+	for i := new(big.Int).Set(x); comp(i, y); i = iter(i) {
+		if bctx.Cancel != nil && bctx.Cancel.Cancelled() {
+			return nil, haltErr
 		}
+		fmt.Println(i)
```
[the remark this answers is not in the document]
Comment (sspaink) [reply topdown/numbers.go]: good catch! removed.

File: topdown/numbers.go [hunk 3]
```diff
@@ -28,32 +28,72 @@ func builtinNumbersRange(bctx BuiltinContext, operands []*ast.Term, iter func(*a
...
+
+	ast, err := builtinRange(bctx, x, y, step, "numbers.range_step")
+	if err != nil {
+		return err
+	}
+
+	return iter(ast)
+}
+
+func builtinRange(bctx BuiltinContext, x *big.Int, y *big.Int, step *big.Int, funcName string) (*ast.Term, error) {
```
[the remark this answers is not in the document]
Comment (sspaink) [reply topdown/numbers.go]: I renamed it to `generateRange` is that better?

File: ast/builtins.go [hunk 4]
```diff
@@ -1347,6 +1348,19 @@ var NumbersRange = &Builtin{
 	),
 }
 
+var NumbersRangeStep = &Builtin{
+	Name:        "numbers.range_step",
+	Description: "Returns an array of numbers in the given (inclusive) range incremented by a positive step. If `a==b`, then `range == [a]`; if `a > b`, then `range` is in descending order.",
```
[the remark this answers is not in the document]
Comment (sspaink) [reply ast/builtins.go]: I added a testcase for this situation and updated the description to mention it. What happens is `b` doesn't get included in the result.

File: test/cases/testdata/numbersrangestep/test-numbersrangestep.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+- note: numbersrangestep/negative
+  query: data.test.p = x
+  modules:
+  - |
+    package test
+
+    p = num {
+      num := numbers.range_step(0, 10, -2)
+    }
+  want_error: 'numbers.range_step: step must be a positive number above zero'
```
[the remark this answers is not in the document]
Comment (sspaink) [reply test/cases/testdata/numbersrangestep/test-numbersrangestep.yaml:45]: Do you think not allowing a negative step is ok?

[the remark this answers is not in the document]
Comment (johanfylling) [reply test/cases/testdata/numbersrangestep/test-numbersrangestep.yaml:45]: I think it'd be ok to allow a negative step; but I think I prefer what you have done here, as it limits the number of permutations for users. If anyone has `a > b`AND a negative `step`, odds are they're making things more complicated for themselves than necessary 😄 . And should this be the wrong assumption, it wouldn't be a breaking change to loosen this restriction in a future release; whereas the reverse would be.

We should definitely mention this imposed restriction in the description, though.

File: topdown/numbers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanfylling) [reply topdown/numbers.go]: Yep, that's great 👍

File: test/cases/testdata/numbersrangestep/test-numbersrangestep.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (sspaink) [reply test/cases/testdata/numbersrangestep/test-numbersrangestep.yaml:45]: Awesome, that makes sense to me! I've updated the description to mention this restrictions: https://github.com/open-policy-agent/opa/pull/6187/files#diff-4b25b18534c046bc03bdda9578775f5bdabad5a3b06001a05f3aed19e0ee4c99R1353-R1356

Let me know if the wording needs any changing.

**Review by johanfylling — APPROVED** [review]
Great work! 👍 
Thanks again for the contribution @sspaink 😃

**Pull request merged**